enterprise/server/githubapp/githubapp.go

```diff
@@ -612,92 +616,136 @@ func (a *GitHubApp) CreateRepo(ctx context.Context, req *rppb.CreateRepoRequest)
 	// Make a temporary directory for the template
-	tmpDir, err := scratchspace.MkdirTemp(tmpDirName)
+	templateTmpDir, err := scratchspace.MkdirTemp(tmpDirName + "-template")
 	if err != nil {
 		return err
 	}
-	defer os.RemoveAll(tmpDir)
+	defer os.RemoveAll(templateTmpDir)
 
 	// Clone the template into the directory
 	auth := &githttp.BasicAuth{
 		Username: "github",
 		Password: token,
 	}
-
-	_, err = git.PlainClone(tmpDir, false, &git.CloneOptions{
+	_, err = git.PlainClone(templateTmpDir, false, &git.CloneOptions{
 		URL:  srcURL,
 		Auth: auth,
 	})
 	if err != nil {
 		return err
 	}
 	// Remove existing git information so we can create a single initial commit
-	err = os.RemoveAll(filepath.Join(tmpDir, ".git"))
+	err = os.RemoveAll(filepath.Join(templateTmpDir, ".git"))
 	if err != nil {
 		return err
 	}
 	// Remove any github workflows directories, because we won't have permission to create them
-	err = os.RemoveAll(filepath.Join(tmpDir, ".github/workflows"))
+	err = os.RemoveAll(filepath.Join(templateTmpDir, ".github/workflows"))
 	if err != nil {
 		return err
 	}
-	// Don't allow paths with dots or other non alphanumeric path components
+	// Don't allow template paths with dots or other non alphanumeric path components
 	if !validPathRegex.MatchString(srcDir) {
 		return status.FailedPreconditionErrorf("invalid template path: %q", srcDir)
 	}
-	path := filepath.Join(tmpDir, srcDir)
+	path := filepath.Join(templateTmpDir, srcDir)
 	// Intentionally using Lstat to avoid following symlinks
 	if fileInfo, err := os.Lstat(path); err != nil || !fileInfo.IsDir() {
 		return status.FailedPreconditionErrorf("not a valid directory: %q", srcDir)
 	}
-	gitRepo, err := git.PlainInit(path, false)
+	// Replace any template variables
+	if len(templateVariables) > 0 {
+		replace(path, templateVariables)
+	}
+	// Make a temporary directory for the new repo
+	repoTmpDir, err := scratchspace.MkdirTemp(tmpDirName + "-repo")
+	if err != nil {
+		return err
+	}
+	defer os.RemoveAll(repoTmpDir)
+
+	newPath := repoTmpDir
+
+	// If we have a destination directory that's not the repo root, update the path
+	if destDir != "" {
+		// Don't allow destinationPath paths with dots or other non alphanumeric path components
+		if !validPathRegex.MatchString(destDir) {
+			return status.FailedPreconditionErrorf("invalid destination path: %q", destDir)
+		}
+		if err := os.RemoveAll(newPath); err != nil {
+			return err
+		}
+		newPath = filepath.Join(repoTmpDir, destDir)
+		if err := os.MkdirAll(newPath, os.ModePerm); err != nil {
+			return err
+		}
+	}
+	// Initialize or clone the destination repo
+	gitRepo, err := initOrClone(needsInit, destDir, destURL, auth)
 	if err != nil {
 		return err
 	}
-	err = gitRepo.Storer.SetReference(plumbing.NewSymbolicReference(plumbing.HEAD, plumbing.Main))
+	fileInfo, err := ioutil.ReadDir(path)
 	if err != nil {
 		return err
 	}
+	// Copy template files into the destination repo
```

Comment:
nit: os.Rename does move (not copy), right?

Reply:
Fixed